pkg/gds/models/v1/contacts.go

```diff
@@ -0,0 +1,202 @@
+package models
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	pb "github.com/trisacrypto/trisa/pkg/trisa/gds/models/v1beta1"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	TechnicalContact      = "technical"
+	AdministrativeContact = "administrative"
+	LegalContact          = "legal"
+	BillingContact        = "billing"
+)
+
+type contactType struct {
+	contact *pb.Contact
+	kind    string
+}
+
+// Returns True if a Contact is not nil and has an email address.
+func ContactHasEmail(contact *pb.Contact) bool {
+	return contact != nil && contact.Email != ""
+}
+
+// Returns True if a Contact is verified.
+func ContactVerified(contact *pb.Contact) (verified bool, err error) {
+	if _, verified, err = GetContactVerification(contact); err != nil {
+		return false, err
+	}
+	return verified, nil
+}
+
+// Returns a standardized order of iterating through contacts.
+func contactOrder(contacts *pb.Contacts) []*contactType {
+	return []*contactType{
+		{contact: contacts.Technical, kind: TechnicalContact},
+		{contact: contacts.Administrative, kind: AdministrativeContact},
+		{contact: contacts.Legal, kind: LegalContact},
+		{contact: contacts.Billing, kind: BillingContact},
+	}
+}
+
+// Returns a function which iterates over the verified contacts in a Contacts object.
+func IterContacts(contacts *pb.Contacts, requireEmail bool) func() (*pb.Contact, string) {
+	all := contactOrder(contacts)
+	i := 0
+	return func() (contact *pb.Contact, kind string) {
+		// Return the next contact that exists or nil to indicate end of iteration.
+		for i < len(all) {
+			contact = all[i].contact
+			kind = all[i].kind
+			i++
+			if contact != nil && (!requireEmail || ContactHasEmail(contact)) {
+				return contact, kind
+			}
+		}
+		return nil, ""
+	}
+}
+
+// Returns a function which iterates over the verified contacts in a Contacts object.
+func IterVerifiedContacts(contacts *pb.Contacts) func() (*pb.Contact, string, error) {
```

Comment:
I'm a little uncomfortable with the error return here, so I'd like to discuss it. The loop terminal condition is when contact is `nil` - but that's true when an error is returned; so will the caller have the opportunity to handle an error? 

E.g. if the first contact errors:

```go
for contact, _, err := contacts(); contact != nil; contact, _ err = contacts() {
    // I don't think err will ever print, will it? 
   fmt.Println(err)
}
``

I think this is why the `iter.Next()` syntax works the way it does; to decouple the loop termination context from error handling. 

Moreover, as mentioned in later comments below - this iterator will not allow us to ignore errors; e.g. if one contact errors but the rest are fine, we still may want to send emails or count validation. What do you think?

Reply:
I see what you mean, I think in order to catch the error the caller has to declare the error outside the loop and check the error value after loop termination, which seems like a lot of work for the caller to do. I think the real solution is to replace the closure with an iterator object with Next() and Value(), this allows us to better distinguish between iterator and value errors and allows the errors to be handled independently and matches the `DirectoryIterator` pattern.